File: app/src/test/java/com/dpis/module/AppListFilterTest.java
package com.dpis.module;

import com.dpis.module.applist.AppListFilterState;

import org.junit.Test;

import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;

public class AppListFilterTest {
    @Test
    public void allAppsTabMatchesBothUserAndSystemApps() {
        assertTrue(matches("",
                AppListFilter.Tab.ALL_APPS,
                "Coolapk",
                "com.coolapk.market",
                false,
                false,
                null,
                null,
                FontApplyMode.OFF,
                null));
        assertTrue(matches("",
                AppListFilter.Tab.ALL_APPS,
                "Android System WebView",
                "com.google.android.webview",
                true,
                false,
                null,
                null,
                FontApplyMode.OFF,
                null));
    }

    @Test
    public void configuredTabUsesConfiguredFlagOnly() {
        assertFalse(matches("",
                AppListFilter.Tab.CONFIGURED_APPS,
                "Coolapk",
                "com.coolapk.market",
                false,
                true,
                null,
                null,
                FontApplyMode.OFF,
                null));
        assertFalse(matches("",
                AppListFilter.Tab.CONFIGURED_APPS,
                "Xiaoheihe",
                "com.max.xiaoheihe",
                false,
                false,
                300,
                null,
                FontApplyMode.OFF,
                null));
        assertFalse(matches("",
                AppListFilter.Tab.CONFIGURED_APPS,
                "Tieba",
                "com.baidu.tieba",
                false,
                false,
                null,
                115,
                FontApplyMode.SYSTEM_EMULATION,
                null));
        assertTrue(matches("",
                AppListFilter.Tab.CONFIGURED_APPS,
                "Tieba",
                "com.baidu.tieba",
                false,
                false,
                null,
                115,
                FontApplyMode.SYSTEM_EMULATION,
                null,
                false,
                true,
                true,
                AppListFilterState.noAdditionalConstraints()));
        assertFalse(matches("",
                AppListFilter.Tab.CONFIGURED_APPS,
                "AdClose",
                "com.close.hook.ads",
                false,
                false,
                null,
                null,
                FontApplyMode.OFF,
                null));
    }

    @Test
    public void configuredTabAndFontOnlyFilterIncludeTypefaceOnlyApps() {
        AppListFilterState fontOnlyState = new AppListFilterState(true, false, false, true);

        assertFalse(matches("",
                AppListFilter.Tab.CONFIGURED_APPS,
                "Reader",
                "com.example.reader",
                false,
                false,
                null,
                null,
                FontApplyMode.OFF,
                "font_abcd1234"));
        assertTrue(matches("",
                AppListFilter.Tab.ALL_APPS,
                "Reader",
                "com.example.reader",
                false,
                false,
                null,
                null,
                FontApplyMode.OFF,
                "font_abcd1234",
                fontOnlyState));
    }

    @Test
    public void configuredTabIncludesAppSpecificConfigOnlyApps() {
        assertFalse(matches("",
                AppListFilter.Tab.CONFIGURED_APPS,
                "WeChat",
                "com.tencent.mm",
                false,
                false,
                null,
                null,
                FontApplyMode.OFF,
                null,
                true,
                AppListFilterState.noAdditionalConstraints()));
        assertTrue(matches("",
                AppListFilter.Tab.CONFIGURED_APPS,
                "WeChat",
                "com.tencent.mm",
                false,
                false,
                null,
                null,
                FontApplyMode.OFF,
                null,
                true,
                true,
                true,
                AppListFilterState.noAdditionalConstraints()));
    }

    @Test
    public void configuredTabIncludesModeOnlyConfiguredApps() {
        assertTrue(matches("",
                AppListFilter.Tab.CONFIGURED_APPS,
                "Mode Only",
                "com.example.mode",
                false,
                false,
                null,
                null,
                FontApplyMode.SYSTEM_EMULATION,
                null,
                false,
                true,
                true,
                AppListFilterState.noAdditionalConstraints()));
    }

    @Test
    public void configuredTabStillSupportsQueryFiltering() {
        assertTrue(matches("tie",
                AppListFilter.Tab.CONFIGURED_APPS,
                "Tieba",
                "com.baidu.tieba",
                false,
                false,
                null,
                115,
                FontApplyMode.SYSTEM_EMULATION,
                null,
                false,
                true,
                true,
                AppListFilterState.noAdditionalConstraints()));
        assertTrue(matches("android",
                AppListFilter.Tab.CONFIGURED_APPS,
                "Android System WebView",
                "com.google.android.webview",
                true,
                true,
                null,
                null,
                FontApplyMode.OFF,
                null,
                false,
                true,
                true,
                AppListFilterState.noAdditionalConstraints()));
        assertFalse(matches("cool",
                AppListFilter.Tab.CONFIGURED_APPS,
                "Android System WebView",
                "com.google.android.webview",
                false,
                true,
                null,
                null,
                FontApplyMode.OFF,
                null,
                false,
                true,
                true,
                AppListFilterState.noAdditionalConstraints()));
    }

    @Test
    public void advancedFiltersCanHideSystemAppsAndRequireInjectedWidthConfig() {
        AppListFilterState state = new AppListFilterState(false, true, true, false);

        assertTrue(matches("",
                AppListFilter.Tab.ALL_APPS,
                "Coolapk",
                "com.coolapk.market",
                false,
                true,
                360,
                null,
                FontApplyMode.OFF,
                null,
                state));
        assertFalse(matches("",
                AppListFilter.Tab.ALL_APPS,
                "Android System WebView",
                "com.google.android.webview",
                true,
                true,
                360,
                null,
                FontApplyMode.OFF,
                null,
                state));
        assertFalse(matches("",
                AppListFilter.Tab.ALL_APPS,
                "Coolapk",
                "com.coolapk.market",
                false,
                false,
                360,
                null,
                FontApplyMode.OFF,
                null,
                state));
        assertFalse(matches("",
                AppListFilter.Tab.ALL_APPS,
                "Coolapk",
                "com.coolapk.market",
                false,
                true,
                null,
                null,
                FontApplyMode.OFF,
                null,
                state));
    }

    @Test
    public void advancedFiltersCanRequireEnabledFontConfig() {
        AppListFilterState state = new AppListFilterState(true, false, false, true);

        assertTrue(matches("",
                AppListFilter.Tab.ALL_APPS,
                "Tieba",
                "com.baidu.tieba",
                false,
                false,
                null,
                115,
                FontApplyMode.SYSTEM_EMULATION,
                null,
                state));
        assertFalse(matches("",
                AppListFilter.Tab.ALL_APPS,
                "Tieba",
                "com.baidu.tieba",
                false,
                false,
                null,
                115,
                FontApplyMode.OFF,
                null,
                state));
        assertFalse(matches("",
                AppListFilter.Tab.ALL_APPS,
                "Tieba",
                "com.baidu.tieba",
                false,
                false,
                null,
                null,
                FontApplyMode.SYSTEM_EMULATION,
                null,
                state));
    }

    @Test
    public void defaultFilterStateStartsWithEveryToggleOff() {
        AppListFilterState state = AppListFilterState.defaultState();

        assertFalse(state.showSystemApps());
        assertFalse(state.injectedOnly());
        assertFalse(state.widthConfiguredOnly());
        assertFalse(state.fontConfiguredOnly());
    }

    private static boolean matches(String query,
            AppListFilter.Tab tab,
            String label,
            String packageName,
            boolean systemApp,
            boolean inScope,
            Integer viewportWidthDp,
            Integer fontScalePercent,
            String fontMode,
            String typefaceId) {
        return matches(query, tab, label, packageName, systemApp, inScope,
                viewportWidthDp, fontScalePercent, fontMode, typefaceId,
                false, false, true, AppListFilterState.noAdditionalConstraints());
    }

    private static boolean matches(String query,
            AppListFilter.Tab tab,
            String label,
            String packageName,
            boolean systemApp,
            boolean inScope,
            Integer viewportWidthDp,
            Integer fontScalePercent,
            String fontMode,
            String typefaceId,
            AppListFilterState state) {
        return matches(query, tab, label, packageName, systemApp, inScope,
                viewportWidthDp, fontScalePercent, fontMode, typefaceId,
                false, false, true, state);
    }

    private static boolean matches(String query,
            AppListFilter.Tab tab,
            String label,
            String packageName,
            boolean systemApp,
            boolean inScope,
            Integer viewportWidthDp,
            Integer fontScalePercent,
            String fontMode,
            String typefaceId,
            boolean appSpecificConfigActive,
            AppListFilterState state) {
        return matches(query, tab, label, packageName, systemApp, inScope,
                viewportWidthDp, fontScalePercent, fontMode, typefaceId,
                appSpecificConfigActive, false, true, state);
    }

    private static boolean matches(String query,
            AppListFilter.Tab tab,
            String label,
            String packageName,
            boolean systemApp,
            boolean inScope,
            Integer viewportWidthDp,
            Integer fontScalePercent,
            String fontMode,
            String typefaceId,
            boolean appSpecificConfigActive,
            boolean configured,
            boolean installed,
            AppListFilterState state) {
        return AppListFilter.matches(query, tab, label, packageName, systemApp, inScope,
                viewportWidthDp, fontScalePercent, fontMode, typefaceId,
                appSpecificConfigActive, configured, installed, state);
    }
}

File: app/src/test/java/com/dpis/module/applist/AppListFilterTest.kt
package com.dpis.module

import com.dpis.module.applist.AppListFilter
import com.dpis.module.applist.AppListFilterState
import com.dpis.module.fonts.FontApplyMode
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

class AppListFilterTest {
    @Test
    fun allAppsTabMatchesBothUserAndSystemApps() {
        assertTrue(matches(label = "Coolapk", packageName = "com.coolapk.market"))
        assertTrue(
            matches(
                label = "Android System WebView",
                packageName = "com.google.android.webview",
                systemApp = true,
            ),
        )
    }

    @Test
    fun configuredTabUsesConfiguredFlagOnly() {
        assertFalse(matches(AppListFilter.Tab.CONFIGURED_APPS, label = "Coolapk", inScope = true))
        assertFalse(
            matches(
                AppListFilter.Tab.CONFIGURED_APPS,
                label = "Xiaoheihe",
                viewportWidthDp = 300
            )
        )
        assertFalse(
            matches(
                AppListFilter.Tab.CONFIGURED_APPS,
                label = "Tieba",
                fontScalePercent = 115,
                fontMode = FontApplyMode.SYSTEM_EMULATION,
            ),
        )
        assertTrue(
            matches(
                AppListFilter.Tab.CONFIGURED_APPS,
                label = "Tieba",
                fontScalePercent = 115,
                fontMode = FontApplyMode.SYSTEM_EMULATION,
                configured = true,
            ),
        )
        assertFalse(matches(AppListFilter.Tab.CONFIGURED_APPS, label = "AdClose"))
    }

    @Test
    fun configuredTabAndFontOnlyFilterIncludeTypefaceOnlyApps() {
        val state = state(configuration = setOf(AppListFilterState.ConfigurationFilter.FONT))

        assertFalse(
            matches(
                AppListFilter.Tab.CONFIGURED_APPS,
                label = "Reader",
                typefaceId = "font_abcd1234"
            )
        )
        assertTrue(
            matches(
                label = "Reader",
                typefaceId = "font_abcd1234",
                state = state,
            ),
        )
    }

    @Test
    fun configuredTabIncludesAppSpecificConfigOnlyApps() {
        assertFalse(
            matches(
                AppListFilter.Tab.CONFIGURED_APPS,
                label = "WeChat",
                appSpecificConfigActive = true,
            ),
        )
        assertTrue(
            matches(
                AppListFilter.Tab.CONFIGURED_APPS,
                label = "WeChat",
                appSpecificConfigActive = true,
                configured = true,
            ),
        )
    }

    @Test
    fun configuredTabIncludesModeOnlyConfiguredApps() {
        assertTrue(
            matches(
                AppListFilter.Tab.CONFIGURED_APPS,
                label = "Mode Only",
                fontMode = FontApplyMode.SYSTEM_EMULATION,
                configured = true,
            ),
        )
    }

    @Test
    fun configuredTabStillSupportsQueryFiltering() {
        assertTrue(
            matches(
                AppListFilter.Tab.CONFIGURED_APPS,
                query = "tie",
                label = "Tieba",
                packageName = "com.baidu.tieba",
                fontScalePercent = 115,
                fontMode = FontApplyMode.SYSTEM_EMULATION,
                configured = true,
            ),
        )
        assertTrue(
            matches(
                AppListFilter.Tab.CONFIGURED_APPS,
                query = "android",
                label = "Android System WebView",
                packageName = "com.google.android.webview",
                inScope = true,
                configured = true,
            ),
        )
        assertFalse(
            matches(
                AppListFilter.Tab.CONFIGURED_APPS,
                query = "cool",
                label = "Android System WebView",
                inScope = true,
                configured = true,
            ),
        )
    }

    @Test
    fun selectedConfigurationFiltersIncludeAnyMatchingCapability() {
        val state = state(
            appTypes = setOf(AppListFilterState.AppType.USER),
            configuration = setOf(
                AppListFilterState.ConfigurationFilter.INJECTED,
                AppListFilterState.ConfigurationFilter.VIEWPORT,
            ),
        )

        assertTrue(matches(label = "Coolapk", inScope = true, viewportWidthDp = 360, state = state))
        assertFalse(
            matches(
                label = "Android System WebView",
                systemApp = true,
                inScope = true,
                viewportWidthDp = 360,
                state = state,
            ),
        )
        assertTrue(matches(label = "Coolapk", viewportWidthDp = 360, state = state))
        assertFalse(matches(label = "Coolapk", state = state))
    }

    @Test
    fun advancedFiltersCanRequireEnabledFontConfig() {
        val state = state(configuration = setOf(AppListFilterState.ConfigurationFilter.FONT))

        assertTrue(
            matches(
                label = "Tieba",
                fontScalePercent = 115,
                fontMode = FontApplyMode.SYSTEM_EMULATION,
                state = state
            ),
        )
        assertFalse(matches(label = "Tieba", fontScalePercent = 115, state = state))
        assertFalse(
            matches(
                label = "Tieba",
                fontMode = FontApplyMode.SYSTEM_EMULATION,
                state = state
            )
        )
    }

    @Test
    fun defaultFilterStateStartsWithEveryToggleOff() {
        val state = AppListFilterState.defaultState()

        assertTrue(state.allAppsSelected())
        assertFalse(state.userAppsSelected())
        assertFalse(state.systemAppsSelected())
        assertFalse(state.injectedOnly())
        assertFalse(state.widthConfiguredOnly())
        assertFalse(state.fontConfiguredOnly())
    }

    @Test
    fun selectedUserAndSystemTypesUseOrSemantics() {
        val state = state(
            appTypes = setOf(AppListFilterState.AppType.USER, AppListFilterState.AppType.SYSTEM),
        )

        assertTrue(matches(label = "User", state = state))
        assertTrue(matches(label = "System", systemApp = true, state = state))
    }

    @Test
    fun selectedConfigurationFiltersUseOrSemantics() {
        val state = state(
            appTypes = setOf(AppListFilterState.AppType.USER),
            configuration = setOf(
                AppListFilterState.ConfigurationFilter.INJECTED,
                AppListFilterState.ConfigurationFilter.FONT,
            ),
        )

        assertTrue(
            matches(
                label = "Reader",
                inScope = true,
                fontScalePercent = 115,
                fontMode = FontApplyMode.SYSTEM_EMULATION,
                state = state
            )
        )
        assertTrue(
            matches(
                label = "Reader",
                fontScalePercent = 115,
                fontMode = FontApplyMode.SYSTEM_EMULATION,
                state = state
            )
        )
        assertTrue(matches(label = "Reader", inScope = true, state = state))
        assertFalse(matches(label = "Reader", state = state))
    }

    @Test
    fun queryMatchesLabelOrPackageNameByContains() {
        assertTrue(matches(query = "ead", label = "Reader", packageName = "com.example.reader"))
        assertTrue(matches(query = "EXAMPLE", label = "Reader", packageName = "com.example.reader"))
        assertFalse(matches(query = "wechat", label = "Reader", packageName = "com.example.reader"))
    }

    private fun state(
        appTypes: Set<AppListFilterState.AppType> = emptySet(),
        configuration: Set<AppListFilterState.ConfigurationFilter> = emptySet(),
    ) = AppListFilterState(appTypes, configuration, AppListFilterState.SortOrder.NAME, false)

    private fun matches(
        tab: AppListFilter.Tab = AppListFilter.Tab.ALL_APPS,
        query: String = "",
        label: String = "Reader",
        packageName: String = "com.example.reader",
        systemApp: Boolean = false,
        inScope: Boolean = false,
        viewportWidthDp: Int? = null,
        fontScalePercent: Int? = null,
        fontMode: String = FontApplyMode.OFF,
        typefaceId: String? = null,
        appSpecificConfigActive: Boolean = false,
        configured: Boolean = false,
        installed: Boolean = true,
        state: AppListFilterState = AppListFilterState.noAdditionalConstraints(),
    ): Boolean = AppListFilter.matches(
        query,
        tab,
        label,
        packageName,
        systemApp,
        inScope,
        viewportWidthDp,
        fontScalePercent,
        fontMode,
        typefaceId,
        null,
        true,
        configured,
        installed,
        state,
    )
}
